driver-core/src/test/unit/com/mongodb/internal/connection/AsynchronousClusterEventListenerTest.java

```diff
@@ -0,0 +1,171 @@
+                new ClusterDescription(ClusterConnectionMode.SINGLE, ClusterType.STANDALONE, Collections.emptyList()));
+        listener.clusterDescriptionChanged(clusterDescriptionChangedEvent);
+        assertEquals(clusterDescriptionChangedEvent, targetListener.take());
+
+        ServerHeartbeatStartedEvent serverHeartbeatStartedEvent = new ServerHeartbeatStartedEvent(connectionId);
+        listener.serverHearbeatStarted(serverHeartbeatStartedEvent);
+        assertEquals(serverHeartbeatStartedEvent, targetListener.take());
+
+        ServerHeartbeatSucceededEvent serverHeartbeatSucceededEvent = new ServerHeartbeatSucceededEvent(connectionId, new BsonDocument(),
+                1, true);
+        listener.serverHeartbeatSucceeded(serverHeartbeatSucceededEvent);
+        assertEquals(serverHeartbeatSucceededEvent, targetListener.take());
+
+        ServerHeartbeatFailedEvent serverHeartbeatFailedEvent = new ServerHeartbeatFailedEvent(connectionId, 1, true, new IOException());
+        listener.serverHeartbeatFailed(serverHeartbeatFailedEvent);
+        assertEquals(serverHeartbeatFailedEvent, targetListener.take());
+
+        ServerOpeningEvent serverOpeningEvent = new ServerOpeningEvent(serverId);
+        listener.serverOpening(serverOpeningEvent);
+        assertEquals(serverOpeningEvent, targetListener.take());
+
+        ServerDescriptionChangedEvent serverDescriptionChangedEvent = new ServerDescriptionChangedEvent(serverId,
+                ServerDescription.builder().address(new ServerAddress()).type(UNKNOWN).state(CONNECTING).build(),
+                ServerDescription.builder().address(new ServerAddress()).type(STANDALONE).state(CONNECTED).build());
+        listener.serverDescriptionChanged(serverDescriptionChangedEvent);
+        assertEquals(serverDescriptionChangedEvent, targetListener.take());
+
+        ServerClosedEvent serverClosedEvent = new ServerClosedEvent(serverId);
+        listener.serverClosed(serverClosedEvent);
+        assertEquals(serverClosedEvent, targetListener.take());
+
+        ClusterClosedEvent clusterClosedEvent = new ClusterClosedEvent(clusterId);
+        listener.clusterClosed(clusterClosedEvent);
+        assertEquals(clusterClosedEvent, targetListener.take());
+
+        // The thread should die after publishing the ClusterClosedEvent
+        listener.getPublishingThread().join(5000);
+    }
+
+    private static final class AllClusterEventListener implements ClusterListener, ServerListener, ServerMonitorListener {
+        private final BlockingQueue<Object> lastEvent = new ArrayBlockingQueue<>(1);
+
+        Object take() throws InterruptedException {
+            return lastEvent.poll(5, TimeUnit.SECONDS);
+        }
+
+        void addEvent(final Object event) {
+            try {
+                lastEvent.put(event);
+            } catch (InterruptedException e) {
+                // ignore
```

Comment:
[optional]
We have not started the test thread, and if it is interrupted, that's likely because the test runner did that. Therefore, we should not erase the information about the thread being interrupted. Preserving this information may be done either by throwing `InterruptedException`, or by setting the interrupted flag back on via `Thread.currentThread().interrupt()`.

Reply:
> This method is called by the listener methods

Ah, sorry :(, it's indeed never called by the thread that runs the test method. In that case, `Assertions.fail()` would have also been an appropriate way to handle the exception.